### PR TITLE
Add loader for `Shader` and `ShaderChunk` asset

### PR DESCRIPTION
**Pull request by Sway007**
### Please check if the PR fulfills these requirements

- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

Add loader for Shader and ShaderChunk asset

## Release Notes

- **New Features**
	- Added a new asset type, `Shader`, to enhance asset categorization.
	- Introduced `ShaderLoader` and `ShaderChunkLoader` classes for improved shader asset management.
	- Added a new utility class `Utils` for URL manipulation and path validation.

- **Improvements**
	- Enhanced resource loading mechanisms across various loaders by integrating `resourceManager` into request methods.
	- Updated `MaterialLoader`, `AnimationClipLoader`, and other loaders for better shader handling and resource management.
	- Improved URL handling in `GLTFBufferParser` and `GLTFSchemaParser` for more robust resource requests.

- **Bug Fixes**
	- Improved error handling for resource retrieval and loading processes in multiple loaders.

- **Deprecations**
	- Marked certain properties in `KTX2Params` and `EngineConfiguration` as deprecated to streamline configurations.

- **Removals**
	- Removed the `MeshLoader` class to simplify resource management.

### CODE REVIEW

File: packages/loader/src/MaterialLoader.ts [hunk 1]
```diff
@@ -32,70 +32,74 @@ class MaterialLoader extends Loader<Material> {
...
-          material.name = name;
+          const { name, shaderData, macros, renderState, shaderRef } = materialSchema;
 
-          const texturePromises = new Array<Promise<Texture2D>>();
-          const materialShaderData = material.shaderData;
-          for (let key in shaderData) {
-            const { type, value } = shaderData[key];
+          // @ts-ignore
+          resourceManager.getResourceByRef<Shader>(<IAssetRef>shaderRef).then((shader) => {
+            const material = new Material(engine, shader);
```
Comment (zhuxudong) [inline packages/loader/src/MaterialLoader.ts:39]: force use shader?pls compatible with common materials and previous materials

Comment (Sway007) [reply packages/loader/src/MaterialLoader.ts]: done

File: packages/core/src/asset/AssetType.ts [hunk 2]
```diff
@@ -23,6 +23,8 @@ export enum AssetType {
   TextureCube = "TextureCube",
   /** Material. */
   Material = "Material",
+  /** Shader */
+  Shader = "Shader",
```
Comment (GuoLei1990) [inline packages/core/src/asset/AssetType.ts:27]: Add '.' in comments

File: packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts [hunk 3]
```diff
@@ -92,3 +92,5 @@ export type IBasicType =
 export type IAssetRef = { key?: string; refId: string };
 
 export type IEntityRef = { entityId: string };
+
+export type IShaderRef = { refId: string };
```
Comment (GuoLei1990) [inline packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts:96]: Use `IAssetRef` is OK?

File: packages/loader/src/ShaderChunkLoader.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  AssetType,
+  LoadItem,
+  Loader,
+  ResourceManager,
+  Shader,
+  ShaderFactory,
+  resourceLoader
+} from "@galacean/engine-core";
+
+@resourceLoader("ShaderChunk", ["glsl"], false)
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:12]: Not cache?

File: packages/loader/src/ShaderChunkLoader.ts [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import {
+  AssetPromise,
+  AssetType,
+  LoadItem,
+  Loader,
+  ResourceManager,
+  Shader,
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:7]: Clear imports

File: packages/loader/src/ShaderChunkLoader.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  ShaderFactory,
+  resourceLoader
+} from "@galacean/engine-core";
+
+@resourceLoader("ShaderChunk", ["glsl"], false)
+class ShaderChunkLoader extends Loader<void> {
+  load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<void> {
+    return this.request<any>(item.url, { ...item, type: "text" }).then(async (code) => {
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:15]: Type is any? Not string?

File: packages/loader/src/ShaderChunkLoader.ts [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+      const { includeKey } = item.params;
+      ShaderFactory.registerInclude(includeKey, code);
+
+      const matches = code.matchAll(/^[ \t]*#include +"([^$\\"]+)"/gm);
+      await Promise.all(
+        Array.from(matches).map((m) => {
+          const path = m[1];
+          if (path) {
+            // @ts-ignore
+            const resource = resourceManager._virtualPathMap[path];
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:25]: Support for import paths can be cleaner, for example, directly associated with absolute URLs. `_virtualPathMap` is just a layer of mapping

File: packages/loader/src/ShaderChunkLoader.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  ShaderFactory,
+  resourceLoader
+} from "@galacean/engine-core";
+
+@resourceLoader("ShaderChunk", ["glsl"], false)
+class ShaderChunkLoader extends Loader<void> {
+  load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<void> {
+    return this.request<any>(item.url, { ...item, type: "text" }).then(async (code) => {
+      const { includeKey } = item.params;
+      ShaderFactory.registerInclude(includeKey, code);
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:17]: You registerInclude with relative path will cause another Shader include same chunk can't find register

File: packages/loader/src/ShaderChunkLoader.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+@resourceLoader("ShaderChunk", ["glsl"], false)
+class ShaderChunkLoader extends Loader<void> {
+  load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<void> {
+    return this.request<any>(item.url, { ...item, type: "text" }).then(async (code: string) => {
+      const { includeKey } = item.params;
+      ShaderFactory.registerInclude(includeKey, code);
+
+      const matches = code.matchAll(/^[ \t]*#include +"([^$\\"]+)"/gm);
+      await Promise.all(
+        Array.from(matches).map((m) => {
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:21]: Why  `Array.from(matches)`? Waste performance and cause GC

File: packages/loader/src/ShaderChunkLoader.ts [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+@resourceLoader("ShaderChunk", ["glsl"], false)
+class ShaderChunkLoader extends Loader<void> {
+  load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<void> {
+    return this.request<any>(item.url, { ...item, type: "text" }).then(async (code: string) => {
+      const { includeKey } = item.params;
+      ShaderFactory.registerInclude(includeKey, code);
+
+      const matches = code.matchAll(/^[ \t]*#include +"([^$\\"]+)"/gm);
+      await Promise.all(
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:20]: Why add await ?

File: packages/loader/src/ShaderChunkLoader.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  ResourceManager,
+  Shader,
+  ShaderFactory,
+  resourceLoader
+} from "@galacean/engine-core";
+
+@resourceLoader("ShaderChunk", ["glsl"], false)
+class ShaderChunkLoader extends Loader<void> {
+  load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<void> {
+    return this.request<any>(item.url, { ...item, type: "text" }).then(async (code: string) => {
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:15]: Why not specify type in generic type?

File: packages/loader/src/MaterialLoader.ts [hunk 12]
```diff
@@ -32,73 +33,91 @@ class MaterialLoader extends Loader<Material> {
...
-                break;
-            }
+          if (shaderRef) {
+            resolve(
+              resourceManager
+                // @ts-ignore
+                .getResourceByRef<Shader>(<IAssetRef>shaderRef)
+                .then((shaderObject) => this.getMaterialByShader(materialSchema, shaderObject, engine))
+            );
+          } else {
```
Comment (zhuxudong) [inline packages/loader/src/MaterialLoader.ts:45]: The builtin shader logic should go here, and should not be repeated in shaderLoader

File: packages/loader/src/ShaderLoader.ts [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+          }
+        })
+      ).then(() => {
+        return Shader.create(code);
+      });
+    });
+  }
+
+  private getBuiltinShader(code: string) {
+    const match = code.match(/^\s*\/\/\s*@builtin\s+(\w+)/);
```
[the remark this answers is not in the document]
Comment (Sway007) [reply packages/loader/src/ShaderLoader.ts]: done

File: packages/loader/src/ShaderChunkLoader.ts (hunk 9 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: cause

File: packages/loader/src/ShaderChunkLoader.ts (hunk 8 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts:10]: Finally! The `includeKey` in params is an absolute path ：）

File: packages/loader/src/ShaderChunkLoader.ts (hunk 6 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: done

File: packages/loader/src/ShaderChunkLoader.ts (hunk 5 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: done

File: packages/loader/src/ShaderChunkLoader.ts (hunk 4 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: done

File: packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts (hunk 3 above)
Comment (Sway007) [reply packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts]: refactored

File: packages/core/src/asset/AssetType.ts (hunk 2 above)
Comment (Sway007) [reply packages/core/src/asset/AssetType.ts:27]: done

File: packages/loader/src/ShaderChunkLoader.ts (hunk 8 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: done

File: packages/loader/src/ShaderChunkLoader.ts (hunk 7 above)
Comment (Sway007) [reply packages/loader/src/ShaderChunkLoader.ts]: All paths are converted to absolute path

File: packages/loader/src/MaterialLoader.ts (hunk 12 above)
Comment (Sway007) [reply packages/loader/src/MaterialLoader.ts:45]: Shaders written with `ShaderLab` can still reference any of builtin shader

**Comment by Sway007** [comment]
after [2df9acb](https://github.com/galacean/engine/pull/2027/commits/2df9acbdb09a74ee34d2a2928255321a21292276), developer can load `Shader` or `ShaderChunk` by http url

File: packages/loader/src/MaterialLoader.ts [hunk 14]
```diff
@@ -32,73 +33,91 @@ class MaterialLoader extends Loader<Material> {
...
-          });
-        })
-        .catch(reject);
+    const texturePromises = new Array<Promise<Texture2D>>();
+    const materialShaderData = material.shaderData;
+    for (let key in shaderData) {
+      const { type, value } = shaderData[key];
+
+      switch (type) {
+        case "Vector2":
```
Comment (zhuxudong) [inline packages/loader/src/MaterialLoader.ts:67]: push these variables into  `EEngineLoadType`, and consume them uniformly in the editor, and delete `transformValues.ts` in editor

File: packages/loader/src/MaterialLoader.ts (hunk 12 above)
Comment (zhuxudong) [reply packages/loader/src/MaterialLoader.ts:45]: * delete 'shaderRef' in editor if is builtinShader
* combine `#include` logic in shader、shaderChunk loader

File: packages/loader/src/MaterialLoader.ts (hunk 14 above)
Comment (Sway007) [reply packages/loader/src/MaterialLoader.ts]: done

**Review by zhuxudong — APPROVED** [review]

File: packages/core/src/asset/Loader.ts [hunk 15]
```diff
@@ -41,5 +42,9 @@ export abstract class Loader<T> {
   constructor(public readonly useCache: boolean) {}
   initialize?(engine: Engine, configuration: EngineConfiguration): Promise<void>;
   abstract load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<T>;
-  request: <U>(url: string, config: RequestConfig) => AssetPromise<U> = request;
+  request<U>(url: string, resourceManager: ResourceManager, config: RequestConfig): AssetPromise<U> {
+    // @ts-ignore
```
Comment (GuoLei1990) [inline packages/core/src/asset/Loader.ts:46]: Why add @ts-ignore

File: packages/core/src/asset/ResourceManager.ts [hunk 16]
```diff
@@ -393,7 +393,7 @@ export class ResourceManager {
   }
 
   private _loadMainAsset<T>(loader: Loader<T>, item: LoadItem, assetBaseURL: string): AssetPromise<T> {
-    item.url = assetBaseURL;
+    // item.url = assetBaseURL;
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:396]: Delete

File: packages/loader/src/GLTFLoader.ts [hunk 17]
```diff
@@ -37,7 +37,8 @@ export class GLTFLoader extends Loader<GLTFResource> {
   }
 
   override load(item: LoadItem, resourceManager: ResourceManager): AssetPromise<GLTFResource> {
-    const url = item.url;
+    // @ts-ignore
+    const url = resourceManager._virtualPathMap[item.url] ?? item.url;
```
Comment (GuoLei1990) [inline packages/loader/src/GLTFLoader.ts:41]: Should resolve use `request<ArrayBuffer>`

File: packages/loader/src/MaterialLoader.ts [hunk 18]
```diff
@@ -26,79 +35,97 @@ function parseProperty(object: Object, key: string, value: any) {
...
-            }
+          if (shaderRef) {
+            resolve(
+              resourceManager
+                // @ts-ignore
+                .getResourceByRef<Shader>(<IAssetRef>shaderRef)
+                .then((shaderObject) => this.getMaterialByShader(materialSchema, shaderObject, engine))
+            );
+          } else {
+            // compatible with 1.2-pre version material schema
```
Comment (GuoLei1990) [inline packages/loader/src/MaterialLoader.ts:54]: Maybe not compatible

File: packages/loader/src/MaterialLoader.ts [hunk 19]
```diff
@@ -26,79 +35,97 @@ function parseProperty(object: Object, key: string, value: any) {
...
+            const shaderObject = Shader.find(shader);
```
Comment (GuoLei1990) [inline packages/loader/src/MaterialLoader.ts:55]: rename

File: packages/loader/src/ShaderChunkLoader.ts [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    return this.request<string>(url, resourceManager, { ...item, type: "text" }).then((code) => {
+      ShaderFactory.registerInclude(url.substring(1), code);
+
+      return _loadChunksInCode(code, url, resourceManager);
+    });
+  }
+}
+
+/** @internal */
+export function _loadChunksInCode(code: string, basePath: string, resourceManager: ResourceManager): Promise<void[]> {
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:27]: Static func is better

File: packages/loader/src/ShaderChunkLoader.ts [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+  const shaderChunkPaths: string[] = [];
+  const matches = code.matchAll(Utils.shaderIncludeRegex);
+  for (const match of matches) {
+    const chunkPath = Utils.pathResolve(match[1], basePath);
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderChunkLoader.ts:31]: Why not use core resolve

File: packages/loader/src/ShaderLoader.ts [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+      }
+
+      return _loadChunksInCode(code, url, resourceManager).then(() => {
+        return Shader.create(code);
+      });
+    });
+  }
+
+  private getBuiltinShader(code: string) {
+    const match = code.match(ShaderLoader._builtinRegex);
```
Comment (GuoLei1990) [inline packages/loader/src/ShaderLoader.ts:32]: _builtinRegex is what?

File: packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts [hunk 23]
```diff
@@ -92,3 +92,5 @@ export type IBasicType =
 export type IAssetRef = { key?: string; refId: string };
 
 export type IEntityRef = { entityId: string };
+
+export type IShaderRef = Omit<IAssetRef, "key">;
```
Comment (GuoLei1990) [inline packages/loader/src/resource-deserialize/resources/schema/BasicSchema.ts:96]: Delete

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**